# **Obtención y preparación de datos**

# OD09. Inspección de Estructuras en Pandas

In [1]:
import pandas as pd
import numpy as np

## <font color='blue'>**Inspección de series y dataframes**</font>

Normalmente, una vez hemos cargado un bloque de datos en una serie o un dataframe, lo primero que haremos será inspeccionarlo para confirmar que los datos cargados son los esperados y que la lectura se ha realizado correctamente. Para esto tenemos los métodos `head`, `tail` y `sample`, con un comportamiento semejante en series y dataframes, que nos muestran un subconjunto de los datos cargados. Además, los métodos `describe` e `info`
 nos proporcionan información adicional sobre los datos.


El método `pandas.Series.head` para series y `pandas.DataFrame.head` para dataframes, devuelve los primeros elementos de la estructura (los primeros valores en el caso de una serie y las primeras filas en el caso de un dataframe). Por defecto, se trata de los 5 primeros elementos, pero podemos especificar el número que deseamos como argumento de la función.

In [2]:
entradas = pd.Series([11, 18, 12, 16, 9, 16, 22, 28, 31, 29, 30, 12],
                     index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"])
entradas

ene    11
feb    18
mar    12
abr    16
may     9
jun    16
jul    22
ago    28
sep    31
oct    29
nov    30
dic    12
dtype: int64

In [3]:
salidas = pd.Series([9, 26, 18, 15, 6, 22, 19, 25, 34, 22, 21, 14],
                    index = ["ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"])
salidas

ene     9
feb    26
mar    18
abr    15
may     6
jun    22
jul    19
ago    25
sep    34
oct    22
nov    21
dic    14
dtype: int64

In [4]:
almacen = pd.DataFrame({"entradas":entradas, "salidas":salidas})
almacen

entradas  salidas
ene        11        9
feb        18       26
mar        12       18
abr        16       15
may         9        6
jun        16       22
jul        22       19
ago        28       25
sep        31       34
oct        29       22
nov        30       21
dic        12       14

In [5]:
almacen = pd.DataFrame({"entradas":entradas, "salidas":salidas})
almacen["neto"] = almacen.entradas - almacen.salidas
almacen

entradas  salidas  neto
ene        11        9     2
feb        18       26    -8
mar        12       18    -6
abr        16       15     1
may         9        6     3
jun        16       22    -6
jul        22       19     3
ago        28       25     3
sep        31       34    -3
oct        29       22     7
nov        30       21     9
dic        12       14    -2

In [6]:
neto = almacen["neto"]
type(neto)

pandas.core.series.Series

En este ejemplo estamos mostrando todos los elementos de la estructura pues son apenas 12. En un caso real podemos estar hablando de miles o de millones de elementos.

Ahora, para mostrar apenas los primeros elementos de la estructura, ejecutamos el método `head`:

In [7]:
entradas.head(7)

ene    11
feb    18
mar    12
abr    16
may     9
jun    16
jul    22
dtype: int64

In [8]:
almacen["neto"].head()

ene    2
feb   -8
mar   -6
abr    1
may    3
Name: neto, dtype: int64

In [9]:
almacen.head()

entradas  salidas  neto
ene        11        9     2
feb        18       26    -8
mar        12       18    -6
abr        16       15     1
may         9        6     3

Los métodos `pandas.Series.tail` (para series) y `pandas.DataFrame.tail` (para dataframes) son semejantes a los anteriores, pero muestran los últimos elementos de la estructura. Si no indicamos otra cosa como argumento, serán los 5 últimos elementos los que se muestren:

In [10]:
entradas.tail()

ago    28
sep    31
oct    29
nov    30
dic    12
dtype: int64

In [11]:
almacen.tail()

entradas  salidas  neto
ago        28       25     3
sep        31       34    -3
oct        29       22     7
nov        30       21     9
dic        12       14    -2

Es frecuente que los datos que hayamos leído estén ordenados según algún criterio, y que el bloque de datos mostrado por los métodos `head` o `tail` estén formados por datos muy parecidos. Y en ocasiones nos puede convenir ver datos aleatorios de nuestra estructura. Para esto podemos utilizar los métodos `pandas.Series.sample` para series y `pandas.DataFrame.sample` para dataframes. Al contrario que `head` o `tail`, el número de elementos devueltos por defecto es uno, por lo que, si deseamos extraer una muestra mayor, tendremos que indicarlo como primer argumento:

In [12]:
entradas.sample(5)

jul    22
jun    16
abr    16
ene    11
mar    12
dtype: int64

In [13]:
almacen.sample(5)

entradas  salidas  neto
jul        22       19     3
oct        29       22     7
ene        11        9     2
ago        28       25     3
abr        16       15     1

El método `describe` devuelve información estadística de los datos del dataframe o de la serie (de hecho, este método devuelve un dataframe). Esta información incluye el número de muestras, el valor medio, la desviación estándar, el valor mínimo, máximo, la mediana y los valores correspondientes a los percentiles 25% y 75%.

Siguiendo con el ejemplo visto en la sección anterior:

In [14]:
almacen.describe()

entradas    salidas       neto
count  12.00000  12.000000  12.000000
mean   19.50000  19.250000   0.250000
std     8.16311   7.641097   5.310795
min     9.00000   6.000000  -8.000000
25%    12.00000  14.750000  -3.750000
50%    17.00000  20.000000   1.500000
75%    28.25000  22.750000   3.000000
max    31.00000  34.000000   9.000000

El método acepta el parámetro `percentiles` conteniendo una lista (o semejante) de los percentiles a mostrar. También acepta los parámetros `include` y `exclude` para espcificar los tipos de las características a incluir o excluir del resultado.

El método `info` muestra un resumen de un dataframe, incluyendo información sobre el tipo de los índices de filas y columnas, los valores no nulos y la memoria usada:

In [15]:
almacen.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12 entries, ene to dic
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   entradas  12 non-null     int64
 1   salidas   12 non-null     int64
 2   neto      12 non-null     int64
dtypes: int64(3)
memory usage: 684.0+ bytes


Solo los dataframes tienen implementado este método.

Un método de las series pandas extremadamente útil es `pandas.Series.value_counts`. Este método devuelve una estructura conteniendo los valores presentes en la serie y el número de ocurrencias de cada uno. Estos valores se muestran en orden decreciente:

In [16]:
s = pd.Series([3, 1, 2, 1, 1, 4, 1, 2, np.nan])
s.value_counts()

1.0    4
2.0    2
3.0    1
4.0    1
Name: count, dtype: int64

Como puede apreciarse, por defecto no se incluyen los valores nulos. Este comportamiento puede modificarse haciendo uso del parámetro `dropna`:

In [17]:
s.value_counts(dropna = False)

1.0    4
2.0    2
3.0    1
4.0    1
NaN    1
Name: count, dtype: int64

En lugar de devolver los valores distintos y el número de ocurrencias, este método también puede agrupar los datos en `bins` y devolver una lista de bins (indicando sus márgenes) con el número de valores en cada uno de ellos. Por ejemplo, si quisiéramos agrupar los valores de la serie anterior en dos bins podríamos hacerlo de la siguiente forma:

In [18]:
s.value_counts(bins = 2)

(0.996, 2.5]    6
(2.5, 4.0]      2
Name: count, dtype: int64

Vemos que se han creados los dos bins, el primero conteniendo los valores entre 0.996 y 2.5 (intervalo abierto por la izquierda y cerrado por la derecha), bin en el que hay 6 valores, y el segundo conteniendo los valores entre 2.5 y 4 (intervalo también abierto por la izquierda y cerrado por la derecha), bin en el que hay 2 valores.

### <font color='green'>Actividad 1</font>

Escribir una función que reciba un diccionario con las alturas (en metros) de los integrantes del grupo y devuelva una serie con la altura mínima, la máxima, media y la desviación estándar.


In [19]:
# Tu código aquí ...

def estadistica_altura(altura):
  datos = pd.Series(altura)
  est= {'min':datos.min(), 'max':datos.max(), 'media':datos.mean(), 'std':datos.std()}
  estadistica = pd.Series(est, name='Estadistica')
  return estadistica

alturas = {'Yisella':1.62, 'Hugo':1.75,'Raul': 1.74,'Begoña' :1.64}
estadistica_altura(alturas)

min      1.62000
max      1.75000
media    1.68750
std      0.06702
Name: Estadistica, dtype: float64

<font color='green'>Fin Actividad 1</font>

### <font color='green'>Actividad 2</font>

Crea un DataFrame a partir del siguiente diccionario y examina las primeras 3 filas y las últimas 2 filas.

```
data = {
    'A': [1, 2, 3, 4, 5],
    'B': [10, 20, 30, 40, 50],
    'C': ['p', 'q', 'r', 's', 't']
}
```



In [20]:
# Tu código aquí ...
data = {
    'A': [1, 2, 3, 4, 5],
    'B': [10, 20, 30, 40, 50],
    'C': ['p', 'q', 'r', 's', 't']
}

df = pd.DataFrame(data)
print('Examinamos las 3 primeras filas')
print(df.head(3))
print()
print('Examinamos las 2 últimas filas')
print(df.tail(2))

Examinamos las 3 primeras filas
   A   B  C
0  1  10  p
1  2  20  q
2  3  30  r

Examinamos las 2 últimas filas
   A   B  C
3  4  40  s
4  5  50  t


<font color='green'>Fin Actividad 2</font>

### <font color='green'>Actividad 3</font>

Utilizando el DataFrame de la actividad anterior, calcular:

* La media de la columna 'A'.
* La suma total de la columna 'B'.
* La descripción estadística del DataFrame.

In [21]:
# Tu código aquí ...
print('La media de la columna A:')
print(df['A'].mean())

print('La suma total de la columna B:')
print(df['B'].sum())

df.describe()

La media de la columna A:
3.0
La suma total de la columna B:
150


A          B
count  5.000000   5.000000
mean   3.000000  30.000000
std    1.581139  15.811388
min    1.000000  10.000000
25%    2.000000  20.000000
50%    3.000000  30.000000
75%    4.000000  40.000000
max    5.000000  50.000000

<font color='green'>Fin Actividad 3</font>

### <font color='green'>Actividad 4</font>

A partir del siguiente DataFrame:

```
df = pd.DataFrame({
    'ID': [101, 102, 103, 104, 105],
    'Value': [23, 45, 12, 67, 34],
    'Category': ['A', 'B', 'A', 'C', 'B']
})
```
* Filtre las filas donde Value sea mayor a 30.
* Ordena el DataFrame por la columna Value en orden descendente.


In [22]:
# Tu código aquí ...
df = pd.DataFrame({
    'ID': [101, 102, 103, 104, 105],
    'Value': [23, 45, 12, 67, 34],
    'Category': ['A', 'B', 'A', 'C', 'B']
})
display(df)
print()

# Filtre las filas donde Value sea mayor a 30.
df30 = df[df['Value']>30]
print(f'Filas del DataFrame donde Value es mayor a 30')
display(df30)
print()

# Ordena el DataFrame por la columna Value en orden descendente.
dfsort = df.sort_values(by='Value', ascending=False)
print(f'DataFrame ordenado segun columna Value')
display(dfsort)

ID  Value Category
0  101     23        A
1  102     45        B
2  103     12        A
3  104     67        C
4  105     34        B


Filas del DataFrame donde Value es mayor a 30


ID  Value Category
1  102     45        B
3  104     67        C
4  105     34        B


DataFrame ordenado segun columna Value


ID  Value Category
3  104     67        C
1  102     45        B
4  105     34        B
0  101     23        A
2  103     12        A

<font color='green'>Fin Actividad 4</font>

### <font color='green'>Actividad 5</font>

Utilizando el DataFrame df de la actividad anterior, crea una nueva columna llamada 'Adjusted', donde cada valor sea el valor original de 'Value' multiplicado por 0.75 si pertenece a la categoría 'A'.

Agrupa el DataFrame por 'Category' y calcula el promedio de las columnas 'Value' y 'Adjusted'.


In [27]:
# Tu código aquí ...
# Tu código aquí ...
df = pd.DataFrame({
    'ID': [101, 102, 103, 104, 105],
    'Value': [23, 45, 12, 67, 34],
    'Category': ['A', 'B', 'A', 'C', 'B']
})
display(df)
display(df.info())
print()
#  Utilizando el DataFrame df de la actividad anterior, crea una nueva columna
#  llamada 'Adjusted', donde cada valor sea el valor original de 'Value'
#  multiplicado por 0.75 si pertenece a la categoría 'A'.

df['Adjusted'] = df['Value'].map(float)
mask = (df['Category']=='A')
df.loc[mask,'Adjusted'] = df.loc[mask,'Value']*0.75
display(df)


#  Agrupa el DataFrame por 'Category' y calcula el promedio de las columnas
#  'Value' y 'Adjusted'.
df_grupo = df.groupby(by='Category')[['Value','Adjusted']].mean()
display(df_grupo)

ID  Value Category
0  101     23        A
1  102     45        B
2  103     12        A
3  104     67        C
4  105     34        B

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ID        5 non-null      int64 
 1   Value     5 non-null      int64 
 2   Category  5 non-null      object
dtypes: int64(2), object(1)
memory usage: 248.0+ bytes


None

ID  Value Category  Adjusted
0  101     23        A     17.25
1  102     45        B     45.00
2  103     12        A      9.00
3  104     67        C     67.00
4  105     34        B     34.00

Value  Adjusted
Category                 
A          17.5    13.125
B          39.5    39.500
C          67.0    67.000

<font color='green'>Fin Actividad 5</font>